<a href="https://colab.research.google.com/github/jmarrietar/LLMs-from-scratch-study/blob/feature%2Fstudy-session/Chapter_2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Es necesario construir primero un diccionaro. El cual va a mapear cada una de las palabras y caracteres especiales a un numero unico. 

#### 2.2 Tokenizing text  

In [2]:
import urllib.request 

url = ("https://raw.githubusercontent.com/rasbt/"
       "LLMs-from-scratch/main/ch02/01_main-chapter-code/"
       "the-verdict.txt")

file_path = "the-verdict.txt"
urllib.request.urlretrieve(url, file_path)

('the-verdict.txt', <http.client.HTTPMessage at 0x7e3fdf4c1a00>)

In [3]:
with open("the-verdict.txt", "r", encoding="utf-8") as f:
    raw_text = f.read()

print("total number of characters:", len(raw_text))
print(raw_text[:99])

total number of characters: 20479
I HAD always thought Jack Gisburn rather a cheap genius--though a good fellow enough--so it was no 


Como dividimos el texto para obtener la lista de tokens? 

In [8]:
# Una forma puede ser utilizando expresiones regulares re. 

# Podemos usar re.split para dividilor en los caracteres en blanco 

import re
text = "Hellow, world. This, is a test"

result = re.split(r"(\s)", text)

print(result)

['Hellow,', ' ', 'world.', ' ', 'This,', ' ', 'is', ' ', 'a', ' ', 'test']


podemos ver que las palabras aun estan conectadas a caracteres especiales de puntuacion. 

In [10]:
# Agregando caracteres especiales. 
result = re.split(r'([,.]|\s)', text)

# Eliminando los caracteres vacios. 
result = [item for item in result if item.strip()]

print(result)

['Hellow', ',', 'world', '.', 'This', ',', 'is', 'a', 'test']


In [11]:
# Otra forma de de tokenizarlo quitando question marks, quotation marks etc. 
text = "Hello, world. Is this-- a test?"
result = re.split(r'([,.:;?_!"()\']|--|\s)', text)
result = [item.strip() for item in result if item.strip()]
print(result)

['Hello', ',', 'world', '.', 'Is', 'this', '--', 'a', 'test', '?']


Aqui ya tendriamos un tokenizador basico

In [13]:
# Ahora aplicarselo a todo el texto. 
preprocessed = re.split(r'([,.:;?_!"()\']|--|\s)', raw_text)
preprocessed = [item.strip() for item in preprocessed if item.strip()]
print(len(preprocessed))


4690


Se tendrian entonces 4690 tokens en todo ese texto

In [16]:
print(preprocessed[:30])

['I', 'HAD', 'always', 'thought', 'Jack', 'Gisburn', 'rather', 'a', 'cheap', 'genius', '--', 'though', 'a', 'good', 'fellow', 'enough', '--', 'so', 'it', 'was', 'no', 'great', 'surprise', 'to', 'me', 'to', 'hear', 'that', ',', 'in']


#### 2.3 Converting tokens into token IDs

Este es el paso intermedio antes de convertirlo en embeddings

Ya todo el texto esta tokenizado en una variable llamada "preprocessed"

In [19]:
# Vamos a ordenar todos los tokens de forma alfabetica

all_words = sorted(set[str](preprocessed))    
vocab_size = len(all_words)
print(vocab_size)

1130


El la lista de tokenns unicos es de 1130 

con estos tokens unicos ya ordenados se puede crear el diccionario. 

In [27]:
vocab = {token:integer for integer, token in enumerate(all_words)}

for i, item in enumerate(vocab.items()):
    print(item)
    if i >=10:
        break

('!', 0)
('"', 1)
("'", 2)
('(', 3)
(')', 4)
(',', 5)
('--', 6)
('.', 7)
(':', 8)
(';', 9)
('?', 10)


¿Que pasa si ya tenemos los token_ids y queremos devolvernos a tener el texto?

Debemos crear un vocabulario inverso que mapee los ids a tokens de texto. 


In [28]:
class SimpleTokenizerV1: 
    def __init__(self, vocab):
        self.str_to_int = vocab # Toca vocabulario
        self.int_to_str = {i:s for s,i in vocab.items()} # creamos un vocab invertido

    def encode(self, text):
        preprocessed = re.split(r'([,.?_!"()\']|--|\s)', text) # split de caracteres 

        # Ahora llevarlos a una lista 
        preprocessed = [item.strip() for item in preprocessed if item.strip()]

        ids = [self.str_to_int[s] for s in preprocessed]

        return ids

    def decode(self, ids): 
        text = " ".join([self.int_to_str[i] for i in ids])

        text = re.sub(r'\s+([,.?!"()\'])', r'\1', text)
        return text 

#### 2.4 Adding special context tokens

#### 2.5 Byte pair encoding

#### 2.6 Data sampling with a sliding window

#### 2.7 Creating token embeddings

#### 2.8 Encoding word position 